From 25182629518c33358eba23fcb6cd3eb005807db9 Mon Sep 17 00:00:00 2001
From: Kazuya Mizuguchi <kazuya.mizuguchi.ks@renesas.com>
Date: Tue, 28 Aug 2018 14:41:06 +0700
Subject: [PATCH 251/909] phy: rcar-gen3-usb2: Add wait for IDDIG

commit be61ce267539e16c9fca8fb793fbf3ef975554bd from
git://git.kernel.org/pub/scm/linux/kernel/git/horms/renesas-bsp.git

According to Figure 73.4 of the R-Car Gen3 Hardware Manual Rev 1.00,
Bit 19: IDDIG of ADPCTRL register needs around 20 ms to be stable.
Adding a delay is necessary to avoid reading illegal value.

Signed-off-by: Kazuya Mizuguchi <kazuya.mizuguchi.ks@renesas.com>
Signed-off-by: An Huynh <an.huynh.uj@renesas.com>
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 drivers/phy/renesas/phy-rcar-gen3-usb2.c | 5 +++++
 1 file changed, 5 insertions(+)

diff --git a/drivers/phy/renesas/phy-rcar-gen3-usb2.c b/drivers/phy/renesas/phy-rcar-gen3-usb2.c
index bb9a37f0297b..1206824158f8 100644
--- a/drivers/phy/renesas/phy-rcar-gen3-usb2.c
+++ b/drivers/phy/renesas/phy-rcar-gen3-usb2.c
@@ -12,6 +12,7 @@
  * published by the Free Software Foundation.
  */
 
+#include <linux/delay.h>
 #include <linux/extcon-provider.h>
 #include <linux/interrupt.h>
 #include <linux/io.h>
@@ -295,6 +296,10 @@ static void rcar_gen3_init_otg(struct rcar_gen3_chan *ch)
 	writel(val | USB2_OBINT_BITS, usb2_base + USB2_OBINTEN);
 	val = readl(usb2_base + USB2_ADPCTRL);
 	writel(val | USB2_ADPCTRL_IDPULLUP, usb2_base + USB2_ADPCTRL);
+
+	/* Wait at least 20 ms for IDDIG status */
+	usleep_range(20000, 21000);
+
 	val = readl(usb2_base + USB2_LINECTRL1);
 	rcar_gen3_set_linectrl(ch, 0, 0);
 	writel(val | USB2_LINECTRL1_DPRPD_EN | USB2_LINECTRL1_DMRPD_EN,
-- 
2.17.1

